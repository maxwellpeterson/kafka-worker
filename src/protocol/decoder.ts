import {
  Int16,
  Int32,
  int16Size,
  int32Size,
  String,
  KafkaArray,
  ErrorCode,
  validErrorCode,
  Int64,
  int64Size,
} from "src/protocol/common";

export class Decoder {
  private readonly view: DataView;
  private offset: number;

  constructor(buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.offset = 0;
  }

  readInt16(): Int16 {
    const value = this.view.getInt16(this.offset);
    this.offset += int16Size;
    return value;
  }

  readEnum<T extends Int16>(predicate: (value: Int16) => value is T): T {
    const value = this.readInt16();
    if (!predicate(value)) {
      throw new Error(`Invalid enum value: ${value}`);
    }
    return value;
  }

  readInt32(): Int32 {
    const value = this.view.getInt32(this.offset);
    this.offset += int32Size;
    return value;
  }

  readInt64(): Int64 {
    const value = this.view.getBigInt64(this.offset);
    this.offset += int64Size;
    return value;
  }

  readString(): String {
    const size = this.readInt16();
    if (size === -1) {
      // I don't think this should ever happen...
      throw new Error("Unexpected null string!");
    }
    const stringBuffer = this.view.buffer.slice(
      this.offset,
      this.offset + size
    );
    const value = new TextDecoder().decode(stringBuffer);
    this.offset += size;
    return value;
  }

  readKafkaArray<T>(readElement: (index: number) => T): KafkaArray<T> {
    const size = this.readInt32();
    if (size === -1) {
      return null;
    }
    const values = new Array<T>(size);
    for (let i = 0; i < size; i++) {
      values[i] = readElement(i);
    }
    return values;
  }

  // In several apis, there is no semantic difference between empty and null
  // arrays, and converting null arrays to empty arrays eliminates unnecessary
  // code pollution from null checking
  readArray<T>(readElement: (index: number) => T): Array<T> {
    return this.readKafkaArray(readElement) ?? [];
  }

  readMessageSet(size: Int32): ArrayBuffer {
    const buffer = this.view.buffer.slice(this.offset, this.offset + size);
    this.offset += size;
    return buffer;
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
