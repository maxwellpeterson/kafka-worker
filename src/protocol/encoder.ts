import {
  Int16,
  Int32,
  int16Size,
  int32Size,
  KafkaString,
  KafkaArray,
} from "src/protocol/common";

export class Encoder {
  private view: DataView;
  private offset: number;

  constructor(buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.offset = int32Size;
  }

  private checkCapacity(size: number) {
    const nextLength = this.offset + size;
    if (nextLength > this.view.byteLength) {
      const newCapacity = 2 * nextLength;
      const newBuffer = new ArrayBuffer(newCapacity);
      // Reserved header space is preserved after copy
      new Uint8Array(newBuffer).set(new Uint8Array(this.view.buffer));
      this.view = new DataView(newBuffer);
    }
  }

  writeInt16(value: Int16) {
    this.checkCapacity(int16Size);
    this.view.setInt16(this.offset, value);
    this.offset += int16Size;
  }

  writeInt32(value: Int32) {
    this.checkCapacity(int32Size);
    this.view.setInt32(this.offset, value);
    this.offset += int32Size;
  }

  writeString(value: KafkaString) {
    if (value === null) {
      this.writeInt16(-1);
      return;
    }
    const bytes = new TextEncoder().encode(value);
    this.writeInt16(bytes.length);
    this.checkCapacity(bytes.length);
    new Uint8Array(this.view.buffer).set(bytes, this.offset);
    this.offset += bytes.length;
  }

  writeArray<T>(values: KafkaArray<T>, writeElement: (value: T) => void) {
    if (values === null) {
      this.writeInt32(-1);
      return;
    }
    this.writeInt32(values.length);
    values.forEach(writeElement);
  }

  sizedBuffer(): ArrayBuffer {
    // Write size header to reserved space at the front of the array
    this.view.setInt32(0, this.offset - int32Size);
    return this.view.buffer.slice(0, this.offset);
  }
}
