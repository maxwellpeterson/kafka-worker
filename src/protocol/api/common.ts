import { Env } from "src/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export type EntrypointFn<Output> = (
  env: Env,
  version: number,
  decoder: Decoder,
  encoder: Encoder
) => Promise<Output>;

export type CheckVersionFn = (version: number) => void;
export type DecodeFn<Req> = (decoder: Decoder) => Req;
export type HandleFn<Req, Res> = (env: Env, request: Req) => Promise<Res>;
export type EncodeFn<Res, Output> = (encoder: Encoder, response: Res) => Output;

export const generateEntrypointFn =
  <Req, Res, Output>(
    apiName: string,
    checkVersion: CheckVersionFn,
    decode: DecodeFn<Req>,
    handle: HandleFn<Req, Res>,
    encode: EncodeFn<Res, Output>
  ): EntrypointFn<Output> =>
  async (
    env: Env,
    version: number,
    decoder: Decoder,
    encoder: Encoder
  ): Promise<Output> => {
    checkVersion(version);

    const request = decode(decoder);
    console.log(
      `Received ${apiName} request: ${JSON.stringify(request, null, 2)}`
    );

    const response = await handle(env, request);
    console.log(
      `Sending ${apiName} response: ${JSON.stringify(response, null, 2)}`
    );

    return encode(encoder, response);
  };

export const generateCheckVersionFn =
  (apiName: string, expected: number): CheckVersionFn =>
  (version: number) => {
    if (version !== expected) {
      throw new Error(
        `Unsupported version of ${apiName} api: expected ${expected} but got ${version}`
      );
    }
  };
