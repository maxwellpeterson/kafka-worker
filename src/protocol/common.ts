import { ValueOf } from "src/common";

// These don't add any type safety, but help with labeling
export type Int16 = number;
export type Int32 = number;
export type Int64 = bigint;

export const int16Size = 2;
export const int32Size = 4;
export const int64Size = 8;

export type KafkaArray<T> = T[] | null;

export interface RequestMetadata {
  apiVersion: Int16;
  correlationId: Int32;
  clientId: string;
}
export const generateEnumPredicate =
  <E extends Record<string, V>, V>(enumObj: E) =>
  (value: V): value is ValueOf<E> =>
    Object.values<V>(enumObj).includes(value);

// ApiKey is an Int16
export const ApiKey = {
  Produce: 0,
  // Fetch: 1,
  // ListOffsets: 2,
  Metadata: 3,
} as const;
export type ApiKey = ValueOf<typeof ApiKey>;
export const validApiKey = generateEnumPredicate(ApiKey);

// ErrorCode is an Int16
export const ErrorCode = {
  None: 0,
  UnknownTopicOrPartition: 3,
} as const;
export type ErrorCode = ValueOf<typeof ErrorCode>;
export const validErrorCode = generateEnumPredicate(ErrorCode);

// Acks is an Int16
export const Acks = {
  None: 0,
  Leader: 1,
  FullISR: -1,
} as const;
export type Acks = ValueOf<typeof Acks>;
export const validAcks = generateEnumPredicate(Acks);
