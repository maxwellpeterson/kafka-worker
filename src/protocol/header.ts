import { Int16, Int32, NullableString } from "src/protocol/common";
import { Decoder, EnumPredicate } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

// Request Header (Version: 1) => api_key api_version correlation_id client_id
//   api_key => INT16
//   api_version => INT16
//   correlation_id => INT32
//   client_id => NULLABLE_STRING
//
// https://kafka.apache.org/protocol.html#protocol_messages
//
// This header format is also used by internal requests.

export interface RequestHeader<T> {
  apiKey: T;
  apiVersion: Int16;
  correlationId: Int32;
  clientId: NullableString;
}

export const encodeRequestHeader = <T extends Int16>(
  encoder: Encoder,
  header: RequestHeader<T>
): Encoder => {
  return encoder
    .writeEnum(header.apiKey)
    .writeInt16(header.apiVersion)
    .writeInt32(header.correlationId)
    .writeNullableString(header.clientId);
};

export const decodeRequestHeader = <T extends Int16>(
  decoder: Decoder,
  predicate: EnumPredicate<T>
): RequestHeader<T> => {
  return {
    apiKey: decoder.readEnum(predicate),
    apiVersion: decoder.readInt16(),
    correlationId: decoder.readInt32(),
    clientId: decoder.readNullableString(),
  };
};

export type RequestMetadata = Omit<RequestHeader<unknown>, "apiKey">;
