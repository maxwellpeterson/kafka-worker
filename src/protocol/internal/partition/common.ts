import { ValueOf } from "src/common";
import {
  ApiKey,
  Int32,
  generateEnumPredicate,
  int32Size,
} from "src/protocol/common";
import { Encoder } from "src/protocol/encoder";

// PartitionApiKey is an Int16
export const PartitionApiKey = {
  Produce: ApiKey.Produce,
} as const;
export type PartitionApiKey = ValueOf<typeof PartitionApiKey>;
export const validPartitionApiKey = generateEnumPredicate(PartitionApiKey);

export class PartitionResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + int32Size);
    this.writeInt32(correlationId);
  }
}
