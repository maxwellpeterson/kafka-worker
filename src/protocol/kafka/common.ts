import { Int32, int32Size } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";

export class KafkaResponseEncoder extends Encoder {
  constructor(correlationId: Int32, initialBufferSize = 64) {
    super(initialBufferSize + 2 * int32Size);
    // Reserve space for size header at front of buffer
    this.offset += int32Size;
    this.writeInt32(correlationId);
  }

  // Overrides parent method, adds size header to buffer
  buffer(): ArrayBuffer {
    // Write size header to reserved space at front of buffer
    this.view.setInt32(0, this.offset - int32Size);
    return super.buffer();
  }
}

export class KafkaRequestDecoder extends Decoder {
  constructor(buffer: ArrayBuffer) {
    super(buffer);

    const expectedSize = this.readInt32();
    const actualSize = buffer.byteLength - int32Size;

    if (expectedSize !== actualSize) {
      throw new Error(
        `Message length does not match size header: expected ${expectedSize} but got ${actualSize}`
      );
    }
  }
}
