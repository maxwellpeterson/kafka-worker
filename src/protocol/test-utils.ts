import { MessageSet } from "src/protocol/common";

// These functions are only intended to be used in test files!

// Convert an ArrayBuffer to a string that can be snapshotted
// Taken from https://stackoverflow.com/a/9458996
export const base64 = (buffer: ArrayBuffer): string => {
  let binary = "";
  const len = buffer.byteLength;
  const bytes = new Uint8Array(buffer);
  for (let i = 0; i < len; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
};

// Generate an arbitrary message set for testing
// TODO: This should return a valid message set!
export const fillMessageSet = (length: number): MessageSet => {
  const buffer = new ArrayBuffer(length);
  const view = new DataView(buffer);
  for (let i = 0; i < length; i++) {
    // This will wrap around when i > 127 but that's fine
    view.setInt8(i, i);
  }
  return new Uint8Array(buffer);
};
