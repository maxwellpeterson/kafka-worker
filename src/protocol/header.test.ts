import { ApiKey, validApiKey } from "src/protocol/common";
import { Decoder } from "src/protocol/decoder";
import { Encoder } from "src/protocol/encoder";
import {
  RequestHeader,
  decodeRequestHeader,
  encodeRequestHeader,
} from "src/protocol/header";
import { base64 } from "./test-utils";

describe("RequestHeader", () => {
  type TestCase = [string, RequestHeader<ApiKey>];
  const cases: TestCase[] = [
    [
      "baseline header",
      {
        apiKey: ApiKey.Metadata,
        apiVersion: 0,
        correlationId: 100,
        clientId: "max",
      },
    ],
    [
      "nonzero api version",
      {
        apiKey: ApiKey.Produce,
        apiVersion: 5,
        correlationId: 25,
        clientId: "max",
      },
    ],
    [
      "empty client id",
      {
        apiKey: ApiKey.Metadata,
        apiVersion: 0,
        correlationId: 1024,
        clientId: "",
      },
    ],
    [
      "null client id",
      {
        apiKey: ApiKey.Metadata,
        apiVersion: 0,
        correlationId: 1024,
        clientId: null,
      },
    ],
  ];

  test.each(cases)("%s", (_name, header) => {
    const encoder = new Encoder();
    const buffer = encodeRequestHeader(encoder, header).buffer();

    expect(base64(buffer)).toMatchSnapshot();

    const decoder = new Decoder(buffer);
    const output = decodeRequestHeader(decoder, validApiKey);

    expect(header).toEqual(output);
  });
});
