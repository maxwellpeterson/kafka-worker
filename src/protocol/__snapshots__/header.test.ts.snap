// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestHeader baseline header 1`] = `"AAMAAAAAAGQAA21heA=="`;

exports[`RequestHeader empty client id 1`] = `"AAMAAAAABAAAAA=="`;

exports[`RequestHeader nonzero api version 1`] = `"AAAABQAAABkAA21heA=="`;

exports[`RequestHeader null client id 1`] = `"AAMAAAAABAD//w=="`;
